### ISSUE TITLE
Design Goals / What have I learned from auto_ml?

### ISSUE THREAD

**Issue by ClimbsRocks** [ClimbsRocks/tabular_ml#2]
What have I learned from auto_ml?

- Keep everything as simple as possible. The main use case is the most important one. If people have enough ML knowledge to want hyper custom stuff, then they should have the tools they need to figure out how to build it for themselves. There are many use cases we just won't be supporting, in order to stay tightly focused on our main use case.
- I hate fat APIs. I really, really want a simple API with a config dict. Again, keep it simple.
- Lol, use pandas. Amazing to think that I built out two different auto_ml projects before I knew how to use pandas.
- I do love good test coverage. TDD is pretty essential for open-source projects to build trust.
- I feel pretty neutral about sklearn's implementation of a pipeline. I love their API design (fit and then transform, etc.), but I seem to remember their actual implementation creating a few issues. 
- Have I mentioned enough to keep it simple? I, of course, want all the bells and whistles. But, that's just not very useable for most users. All the advanced stuff makes sense to me, because I wrote it. But, for most other advanced users, they want their own way of implementing most of these things. I might even try to do something like that intentionally, having a main version of this repo (this one!) and then separately having a highly-customized repo that lets users do all kinds of fun, advanced ML stuff. Maybe something like `tabular_ml_pro`? Doing that would give me an avenue to add in all the advanced ML stuff that I want to, while still having an excuse to keep this version of the project as clean and simple as I possibly can. 
- Maintainer time is expensive. It's really like a seed-stage startup- one of the biggest risks to any early-stage startup is the founders losing interest and motivation. I do like interacting with strangers on the internet around things that I enjoy (like this!) But, it can also be draining. Some initial thoughts on how to address this are included below.

Design goals for tabular_ml

- [ ] Run in both batch and single predictions, at production-ready speed.
- [ ] Give the user robust analytics. Coefficients, feature_importances, feature_responsivenes.
- [ ] Run out of the box with reasonable sets of defaults and best practices, but let the user specify configs for pretty much everything that they want.
- [ ] Clean API. Pretty much everything goes in the configs. It takes in just a df.
- [ ] Let the user specify their own data transform layer. 
- [ ] Probably let the user specify their own model layer? Or, maybe to put it differently, let people use this for just the data transform layers, not the modeling layer.
- [ ] Most stuff after the model is out of scope. Sure, we'll do logging/scoring of results. But everything else (ensembling, etc) is going to be largely out of scope. 
- [ ] Lots of charts. Lots of visuals. We'll need to figure out how (do we have an "interactive" mode, for jupyter notebooks and such, that displays visuals? Do we try to print visuals in the shell? Do we pop up visuals? Do we write them to disk somewhere, and let the user stream that file on disk?)
- [ ] Maybe: a system to track experiments and results.
- [ ] Maybe: hyperparameter tuning. Might be out of scope just due to the increased complexity. People can still do this manually with different configs. It was rarely worth the compute for all the models I've trained in the past. I don't remember it being a widely used feature before.

How to keep maintainer interest/motivation high:
1. Encourage people to star the project! It does signal interest in a way that keeps me engaged.
2. Good documentation.
3. Display personality. If I'm going to do this, I might as well do it my way. 
4. Create a culture. Have an Issue template that requires people to upload an "awww" image with their issues. Encourage people to come up with fun names for their projects. Add bright colors. 
5. Have a clear vision of what the project's main use case is, and be ruthlessly consistent about staying on target to that main mission. Find a place to put the distractions (a whole different repo, for instance, that has all kinds of advanced features).
6. Have good test coverage, and a CI infra built out.

I'm sure there's more to all of these sections, but I'm antsy to get to more code design!
